### PR TITLE
Create `geodesic` module 

### PR DESCRIPTION
**Pull request by diegoabt**
This module contains the implementation of the Geodesic Sinkhorn algorithm [1].

[1] Huguet, G., Tong, A., Zapatero, M. R., Wolf, G., & Krishnaswamy, S. (2022). Geodesic Sinkhorn: optimal transport for high-dimensional datasets. arXiv preprint arXiv:2211.00805.

### CODE REVIEW

**Comment by marcocuturi** [comment]
thanks a lot Diego! Do you think you can add some basic testing? Thanks!

**Comment by diegoabt** [comment]
You're very welcome! Yes, for sure! Do you mean, adding it in the `tests` dir directly?

**Comment by marcocuturi** [comment]
indeed, something like https://github.com/ott-jax/ott/blob/main/tests/geometry/graph_test.py

File: src/ott/geometry/geodesic.py [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+  def apply_kernel(
+      self,
+      scaling: jnp.ndarray,
+      eps: Optional[float] = None,
+      axis: int = 0,
+  ) -> jnp.ndarray:
+    r"""Apply :attr:`kernel_matrix` on positive scaling vector.
+
+    Args:
+    scaling: Scaling to apply the kernel to.
```
Comment (michalk8) [inline src/ott/geometry/geodesic.py:122]: Please correctly dedent.

File: src/ott/geometry/geodesic.py [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+    eps: passed for consistency, not used yet.
+    axis: passed for consistency, not used yet.
+
+    Returns:
+    Kernel applied to ``scaling``.
```
Comment (michalk8) [inline src/ott/geometry/geodesic.py:127]: Please correctly dedent.

File: src/ott/geometry/geodesic.py [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+    """
+
+    def compute_laplacian(adjacency_matrix: jnp.ndarray) -> jnp.ndarray:
```
Comment (michalk8) [inline src/ott/geometry/geodesic.py:130]: As discussed in the `from_graph`, these functions to compute the Chebyshev polynomials, etc. should be done there, as this is executed every Sinkhorn iteration.

File: src/ott/geometry/geodesic.py [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+
+    return compute_chebyshev_approximation(
+        laplacian_times_signal, chebyshev_coeffs
+    )
+
+  @property
+  def kernel_matrix(self) -> jnp.ndarray:  # noqa: D102
+    n, _ = self.shape
+    kernel = self.apply_kernel(jnp.eye(n))
+    # force symmetry because of numerical imprecision
```
Comment (michalk8) [inline src/ott/geometry/geodesic.py:230]: Remove the comment + forced symmetrization.

File: src/ott/geometry/geodesic.py [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+      """Compute the Chebyshev polynomial approx for the given input and coeffs.
+
+      Args:
+          x: Input to evaluate the polynomial at.
+          coeffs: List of Chebyshev polynomial coefficients.
+
+      Returns:
+          The Chebyshev polynomial approximation evaluated at x.
+      """
+      return self.apply_kernel(x, coeffs)
```
Comment (michalk8) [inline src/ott/geometry/geodesic.py:213]: This will cause `RecursionError`.

File: src/ott/geometry/geodesic.py [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+      """Compute the coeffs of the Chebyshev pols approx using Bessel functs.
+
+      Args:
+          t: Time parameter.
+          max_order: Maximum order of the Chebyshev polynomial approximation.
+
+      Returns:
+          A list of coefficients.
+      """
+      return (2 * ive(jnp.arange(0, max_order + 1), -t)).tolist()
```
Comment (michalk8) [inline src/ott/geometry/geodesic.py:199]: This has to be done using https://jax.readthedocs.io/en/latest/_autosummary/jax.pure_callback.html, as you call the `scipy` implementation. Will go check the status of JAX implementation, but afaik it's not ready, hence the need for the pure callback.

File: src/ott/geometry/geodesic.py [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+    kernel = self.apply_kernel(jnp.eye(n))
+    # force symmetry because of numerical imprecision
+    # happens when `numerical_scheme='backward_euler'` and small `t`
+    return (kernel + kernel.T) * 0.5
+
+  @property
+  def cost_matrix(self) -> jnp.ndarray:  # noqa: D102
+    return -self.t * mu.safe_log(self.kernel_matrix)
+
+  @property
```
Comment (michalk8) [inline src/ott/geometry/geodesic.py:238]: Unused, remove please.

File: src/ott/geometry/geodesic.py [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+  def kernel_matrix(self) -> jnp.ndarray:  # noqa: D102
+    n, _ = self.shape
+    kernel = self.apply_kernel(jnp.eye(n))
+    # force symmetry because of numerical imprecision
+    # happens when `numerical_scheme='backward_euler'` and small `t`
+    return (kernel + kernel.T) * 0.5
+
+  @property
+  def cost_matrix(self) -> jnp.ndarray:  # noqa: D102
+    return -self.t * mu.safe_log(self.kernel_matrix)
```
Comment (michalk8) [inline src/ott/geometry/geodesic.py:236]: I'm not sure whether this holds for the `Geodesic` geometry.

File: src/ott/geometry/geodesic.py [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+    if self.numerical_scheme == "backward_euler":
+      return self.t / (4. * self.n_steps)
+    if self.numerical_scheme == "crank_nicolson":
+      return self.t / (2. * self.n_steps)
+    raise NotImplementedError(
+        f"Numerical scheme `{self.numerical_scheme}` is not implemented."
+    )
+
+  @property
+  def _scaled_laplacian(self) -> jnp.ndarray:
```
Comment (michalk8) [inline src/ott/geometry/geodesic.py:250]: Unused.

File: src/ott/geometry/geodesic.py [hunk 10]
```diff
@@ -0,0 +1,308 @@
...
+    """Laplacian scaled by a constant, depending on the numerical scheme."""
+    return self._scale * self.laplacian
+
+  @property
+  def _M(self) -> jnp.ndarray:
```
Comment (michalk8) [inline src/ott/geometry/geodesic.py:255]: Unused.

File: src/ott/geometry/geodesic.py [hunk 11]
```diff
@@ -0,0 +1,308 @@
...
+  def marginal_from_potentials(
+      self,
+      f: jnp.ndarray,
+      g: jnp.ndarray,
+      axis: int = 0,
+  ) -> jnp.ndarray:
+    """Not implemented."""
+    raise ValueError("Not implemented.")
+
+  def tree_flatten(self) -> Tuple[Sequence[Any], Dict[str, Any]]:  # noqa: D102
```
Comment (michalk8) [inline src/ott/geometry/geodesic.py:297]: Please modify this according to `__init__` after you modify/remove the arguments.

File: src/ott/geometry/geodesic.py [hunk 12]
```diff
@@ -0,0 +1,308 @@
...
+          k: Number of eigenvalues/vectors to compute.
+
+      Returns:
+          The largest eigenvalue of the Laplacian matrix.
+      """
+      n = laplacian_matrix.shape[0]
+      initial_directions = jax.random.normal(jax.random.PRNGKey(0), (n, k))
+      # Convert the Laplacian matrix to a dense array
+      #laplacian_array = laplacian_matrix.toarray()
+      eigvals, _, _ = lobpcg_standard(laplacian_matrix, initial_directions, m=k)
```
Comment (michalk8) [inline src/ott/geometry/geodesic.py:157]: `lobpcg_standard` doesn't seem to accept a sparse matrix by default, but you can define the matrix-vector product using as `jax.experimental.sparse.sparsify(lambda v: lap @ v)`. Currently the best supported format is the [BCOO](https://jax.readthedocs.io/en/latest/_autosummary/jax.experimental.sparse.BCOO.html). When refactoring, please start assuming that the laplacian can also be a BCOO sparse matrix.
Here are some further pointers: https://github.com/google/lineax/issues/24

File: src/ott/geometry/geodesic.py [hunk 13]
```diff
@@ -0,0 +1,308 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Any, Dict, List, Optional, Sequence, Tuple
+
+import jax
+import jax.numpy as jnp
+import numpy as np
+from jax.experimental.sparse.linalg import lobpcg_standard
```
Comment (michalk8) [inline src/ott/geometry/geodesic.py:20]: Would use `import jax.experimental.sparse as jesp` and use `jesp.linalg.lobpcg_standard` later.

File: src/ott/geometry/geodesic.py [hunk 14]
```diff
@@ -0,0 +1,308 @@
...
+from scipy.special import ive
+
+from ott.geometry import geometry
+from ott.math import utils as mu
+
+__all__ = ["Geodesic"]
+
+
+@jax.tree_util.register_pytree_node_class
+class Geodesic(geometry.Geometry):
```
Comment (michalk8) [inline src/ott/geometry/geodesic.py:30]: Please add this class to the `docs/geometry.rst`.

File: src/ott/geometry/geodesic.py [hunk 15]
```diff
@@ -0,0 +1,308 @@
...
+          adjacency matrix of a graph.
+
+      Returns:
+          An (n, n) array representing the Laplacian matrix.
+      """
+      degree_matrix = jnp.diag(jnp.sum(adjacency_matrix, axis=0))
+      return degree_matrix - adjacency_matrix
+
+    def compute_largest_eigenvalue(laplacian_matrix, k):
+      """Compute the largest eigenvalue of the Laplacian matrix.
```
Comment (michalk8) [inline src/ott/geometry/geodesic.py:144]: Local and fairly simple function like these don't need to have such comprehensive docstrings; would either move them out the class as helper function or remove/simplify the docstrinsg (having comments might be slightly better).

File: src/ott/geometry/geodesic.py [hunk 16]
```diff
@@ -0,0 +1,308 @@
...
+
+      Args:
+          laplacian_matrix: An (n, n) array representing the Laplacian matrix.
```
Comment (michalk8) [inline src/ott/geometry/geodesic.py:147]: Wrong indentation (should be 2, not 4), also in other places.

File: src/ott/geometry/geodesic.py [hunk 17]
```diff
@@ -0,0 +1,308 @@
...
+          k: Number of eigenvalues/vectors to compute.
+
+      Returns:
+          The largest eigenvalue of the Laplacian matrix.
+      """
+      n = laplacian_matrix.shape[0]
```
Comment (michalk8) [inline src/ott/geometry/geodesic.py:153]: Use `n, _ = self.shape`

File: src/ott/geometry/geodesic.py [hunk 18]
```diff
@@ -0,0 +1,308 @@
...
+      initial_directions = jax.random.normal(jax.random.PRNGKey(0), (n, k))
```
Comment (michalk8) [inline src/ott/geometry/geodesic.py:154]: I wouldn't hardcode the random key; rather in `from_graph`, I'd allow for passing the key or using `ott.utils.default_prng_key`.

File: src/ott/geometry/geodesic.py [hunk 19]
```diff
@@ -0,0 +1,308 @@
...
+
+      Returns:
+          The Chebyshev polynomial approximation evaluated at x.
+      """
+      return self.apply_kernel(x, coeffs)
+
+    #laplacian_matrix = compute_laplacian(self.adjacency_matrix)
+    rescaled_laplacian = rescale_laplacian(self.laplacian)
+    scaled_laplacian = define_scaled_laplacian(rescaled_laplacian)
+    chebyshev_coeffs = chebyshev_coefficients(self.t, self.n_steps)
```
Comment (michalk8) [inline src/ott/geometry/geodesic.py:218]: This shouldn't be really called `n_steps`, but maybe `order` (for the max order of Chebyshev polynomials).

File: src/ott/geometry/geodesic.py [hunk 20]
```diff
@@ -0,0 +1,308 @@
...
+
+from ott.geometry import geometry
+from ott.math import utils as mu
+
+__all__ = ["Geodesic"]
+
+
+@jax.tree_util.register_pytree_node_class
+class Geodesic(geometry.Geometry):
+  r"""Graph distance approximation using heat kernel :cite:`heitz:21,crane:13`.
```
[the remark this answers is not in the document]
Comment (diegoabt) [reply src/ott/geometry/geodesic.py:31]: Added.

File: src/ott/geometry/geodesic.py [hunk 21]
```diff
@@ -0,0 +1,308 @@
...
+
+  Approximates the heat kernel for large ``n_steps``, which for small ``t``
+  approximates the geodesic exponential kernel :math:`e^{\frac{-d(x, y)^2}{t}}`.
+
+  Args:
```
[the remark this answers is not in the document]
Comment (diegoabt) [reply src/ott/geometry/geodesic.py:36]: Updated.

File: src/ott/geometry/geodesic.py [hunk 22]
```diff
@@ -0,0 +1,308 @@
...
+  @classmethod
+  def from_graph(
+      cls,
+      G: jnp.ndarray,
+      t: Optional[float] = 1e-3,
+      directed: bool = False,
+      normalize: bool = False,
+      **kwargs: Any
+  ) -> "Geodesic":
+    r"""Construct :class:`~ott.geometry.graph.Graph` from an adjacency matrix.
```
[the remark this answers is not in the document]
Comment (diegoabt) [reply src/ott/geometry/geodesic.py:74]: Updated.

File: src/ott/geometry/geodesic.py (hunk 4 above)
Comment (diegoabt) [reply src/ott/geometry/geodesic.py:230]: Removed.

File: src/ott/geometry/geodesic.py (hunk 9 above)
Comment (diegoabt) [reply src/ott/geometry/geodesic.py:250]: Removed.

File: src/ott/geometry/geodesic.py (hunk 10 above)
Comment (diegoabt) [reply src/ott/geometry/geodesic.py:255]: Removed.

File: src/ott/geometry/geodesic.py (hunk 17 above)
Comment (diegoabt) [reply src/ott/geometry/geodesic.py:153]: Used.

File: src/ott/geometry/geodesic.py (hunk 19 above)
Comment (diegoabt) [reply src/ott/geometry/geodesic.py:218]: Corrected.

File: src/ott/geometry/geodesic.py (hunk 18 above)
Comment (diegoabt) [reply src/ott/geometry/geodesic.py:144]: Solved.

File: src/ott/geometry/geodesic.py (hunk 15 above)
Comment (diegoabt) [reply src/ott/geometry/geodesic.py:135]: Changed the docstrings by comments so it is easier to read.

File: src/ott/geometry/geodesic.py [hunk 23]
```diff
@@ -0,0 +1,308 @@
...
+      **kwargs: Any
+  ):
+    super().__init__(epsilon=1., **kwargs)
+    self.laplacian = laplacian
+    self.t = t
+    self.n_steps = n_steps
+    self.tol = tol
+
+  @classmethod
+  def from_graph(
```
[the remark this answers is not in the document]
Comment (diegoabt) [reply src/ott/geometry/geodesic.py:60]: Done.

File: src/ott/geometry/geodesic.py (hunk 5 above)
Comment (diegoabt) [reply src/ott/geometry/geodesic.py]: Fixed.

File: src/ott/geometry/geodesic.py (hunk 13 above)
Comment (diegoabt) [reply src/ott/geometry/geodesic.py:21]: Corrected.

File: src/ott/geometry/geodesic.py (hunk 11 above)
Comment (diegoabt) [reply src/ott/geometry/geodesic.py:238]: Fixed.

File: src/ott/geometry/geodesic.py (hunk 12 above)
Comment (diegoabt) [reply src/ott/geometry/geodesic.py]: Done.

File: src/ott/geometry/geodesic.py (hunk 8 above)
Comment (diegoabt) [reply src/ott/geometry/geodesic.py]: Corrected using the Eq. (5) of the [main ref](https://www.semanticscholar.org/reader/eb490736a6cfc0cc2de8a64ea032928b8e2a7c45).

File: src/ott/geometry/geodesic.py (hunk 6 above)
Comment (diegoabt) [reply src/ott/geometry/geodesic.py]: I have included this function into the `from_graph()` method, and it is looking now like this:

`
# Compute the coeffs of the Chebyshev pols approx using Bessel functs.
chebyshev_coeffs = (2 * ive(jnp.arange(0, geodesic_instance.order + 1), -geodesic_instance.t)).tolist()
`

Does it still need to be wrapped around the callback?

File: src/ott/geometry/geodesic.py (hunk 14 above)
Comment (diegoabt) [reply src/ott/geometry/geodesic.py:30]: Added.

File: src/ott/geometry/geodesic.py (hunk 3 above)
Comment (diegoabt) [reply src/ott/geometry/geodesic.py]: Moved.

File: src/ott/geometry/geodesic.py [hunk 24]
```diff
@@ -0,0 +1,260 @@
...
+    self.order = order
+    self.chebyshev_coeffs = chebyshev_coeffs
+    self.numerical_scheme = numerical_scheme
+
+  @classmethod
+  def from_graph(
+      cls,
+      G: jnp.ndarray,
+      t: Optional[float] = 1e-3,
+      order=100,
```
Comment (michalk8) [inline src/ott/geometry/geodesic.py:66]: order: int = 100

File: src/ott/geometry/geodesic.py [hunk 25]
```diff
@@ -0,0 +1,260 @@
...
+    self.laplacian = laplacian
+    self.t = t
+    self.order = order
+    self.chebyshev_coeffs = chebyshev_coeffs
+    self.numerical_scheme = numerical_scheme
+
+  @classmethod
+  def from_graph(
+      cls,
+      G: jnp.ndarray,
```
Comment (michalk8) [inline src/ott/geometry/geodesic.py:64]: Change the type (see above comment).

File: src/ott/geometry/geodesic.py [hunk 26]
```diff
@@ -0,0 +1,260 @@
...
+    axis: passed for consistency, not used yet.
+
+    Returns:
+    Kernel applied to ``scaling``.
+    """
+
+    def compute_largest_eigenvalue(laplacian_matrix, k, seed=None):
+      # Compute the largest eigenvalue of the Laplacian matrix.
+      if seed is None:
+        seed = jax.random.PRNGKey(0)
```
Comment (michalk8) [inline src/ott/geometry/geodesic.py:143]: When refactoring this, use `from ott import utils; utils.default_prng_key`.

File: src/ott/geometry/geodesic.py [hunk 27]
```diff
@@ -0,0 +1,260 @@
...
+      # Create a sparse matrix-vector product function using sparsify
+      # This function multiplies the sparse laplacian_matrix with a vector
+      lapl_vector_product = jesp.sparsify(lambda v: laplacian_matrix @ v)
+
+      # Compute eigenvalues using the sparse matrix-vector product
+      eigvals, _, _ = jesp.linalg.lobpcg_standard(
+          lapl_vector_product, initial_dirs, m=k
+      )
+
+      return np.max(eigvals)
```
Comment (michalk8) [inline src/ott/geometry/geodesic.py:157]: jnp.max

File: src/ott/geometry/geodesic.py [hunk 28]
```diff
@@ -0,0 +1,260 @@
...
+      identity = jnp.eye(n)
+      return laplacian_matrix - identity
+
+    def compute_chebyshev_approximation(
+        x: jnp.ndarray, coeffs: List[float]
+    ) -> jnp.ndarray:
+      # Compute the Chebyshev polynomial approximation for
+      # the given input and coefficients.
+      x_dense = x.toarray() if sp.issparse(x) else x
+      return np.polynomial.chebyshev.chebval(x_dense, coeffs)
```
Comment (michalk8) [inline src/ott/geometry/geodesic.py:180]: Wrap this in `jax.pure_callback`, since it's scipy function.

File: src/ott/geometry/geodesic.py [hunk 29]
```diff
@@ -0,0 +1,260 @@
...
+      n = laplacian_matrix.shape[0]
+      identity = jnp.eye(n)
+      return laplacian_matrix - identity
+
+    def compute_chebyshev_approximation(
+        x: jnp.ndarray, coeffs: List[float]
+    ) -> jnp.ndarray:
+      # Compute the Chebyshev polynomial approximation for
+      # the given input and coefficients.
+      x_dense = x.toarray() if sp.issparse(x) else x
```
Comment (michalk8) [inline src/ott/geometry/geodesic.py:179]: Don't use `scipy` here, assume jax.Array or jesp.BCOO

**Review by michalk8 — CHANGES_REQUESTED** [review]

File: src/ott/geometry/geodesic.py (hunk 27 above)
Comment (diegoabt) [reply src/ott/geometry/geodesic.py:157]: Corrected.

File: src/ott/geometry/geodesic.py (hunk 26 above)
Comment (diegoabt) [reply src/ott/geometry/geodesic.py]: Fixed.

File: src/ott/geometry/geodesic.py [hunk 30]
```diff
@@ -0,0 +1,260 @@
...
+
+  @property
+  def kernel_matrix(self) -> jnp.ndarray:  # noqa: D102
+    n, _ = self.shape
+    return self.apply_kernel(jnp.eye(n))
+
+  @property
+  def cost_matrix(self) -> jnp.ndarray:  # noqa: D102
+    # Calculate the cost matrix using the formula (5) from the main reference
+    return -4 * self.t * jnp.log(self.kernel_matrix)
```
[the remark this answers is not in the document]
Comment (diegoabt) [reply src/ott/geometry/geodesic.py:200]: Corrected

File: src/ott/geometry/geodesic.py (hunk 24 above)
Comment (diegoabt) [reply src/ott/geometry/geodesic.py:68]: Corrected

File: src/ott/geometry/geodesic.py (hunk 29 above)
Comment (diegoabt) [reply src/ott/geometry/geodesic.py:181]: Changed it to jesp.BCOO

File: src/ott/geometry/geodesic.py [hunk 31]
```diff
@@ -0,0 +1,260 @@
...
+      inv_sqrt_deg = jnp.diag(
+          jnp.where(degree > 0.0, 1.0 / jnp.sqrt(degree), 0.0)
+      )
+      laplacian = inv_sqrt_deg @ laplacian @ inv_sqrt_deg
+
+    if t is None:
+      t = (jnp.sum(G) / jnp.sum(G > 0.)) ** 2
+
+    # Create an instance of the Geodesic class and set the attribute
+    geodesic_instance = cls(laplacian, t=t, order=order, **kwargs)
```
[the remark this answers is not in the document]
Comment (diegoabt) [reply src/ott/geometry/geodesic.py]: Corrected

**Comment by michalk8** [comment]
@diegoabt 
I spoke with @MUCDK and @guillaumehu, Guillaume agreed to finish PR (the rest of the open comments, tests, etc.) this in #457 - hope that's alright with you - we wanna include this asap.
I will be closing this in favor of #457, thanks a lot for doing this!

**Comment by diegoabt** [comment]
Hi @michalk8
Sounds good! Hope it's finished soon. Good luck!

**Pull request closed without merging**